integration-tests/tests/src/tests/listeners.ts

```diff
@@ -0,0 +1,173 @@
+    it("should work for change", function (this: RealmContext, done) {
+      this.realm.addListener("change", (realm, name, schema) => {
+        expect(name).to.equal("change");
+        expect(realm).to.equal(this.realm);
+        expect(schema).to.equal(undefined);
+        done();
+      });
+
+      this.realm.write(() => {
+        this.realm.create("Person", { age: 3, name: "Bob" });
+      });
+    });
+    it("should work for beforenotify", function (this: RealmContext, done) {
+      this.realm.addListener("beforenotify", (realm, name, schema) => {
+        expect(name).to.equal("beforenotify");
+        expect(realm).to.equal(this.realm);
+        expect(schema).to.equal(undefined);
+        done();
+      });
+      this.realm.write(() => {
+        this.realm.create("Person", { age: 3, name: "Bob" });
+      });
+    });
+    it("should work for schema", function (this: RealmContext, done) {
+      this.realm.addListener("schema", (realm, name, schema) => {
+        expect(name).to.equal("schema");
+        expect(realm).to.equal(this.realm);
+        expect(schema).to.deep.equal(this.realm.schema);
+        done();
+      });
+
+      this.realm.write(() => {
+        this.realm._updateSchema([...this.realm.schema, { name: "MyClass", properties: { myField: "string" } }]);
+      });
+    });
+    it("should not call change and schema for no reason", function (this: RealmContext, done) {
+      this.realm.addListener("change", () => {
+        done("should not run");
+      });
+      this.realm.addListener("schema", () => {
+        done("should not run");
+      });
+      done();
+    });
+    it("should work with multiple listeners of same type", function (this: RealmContext, done) {
+      let firstListenerRuns = 0;
+      let secondListenerRuns = 0;
+      this.realm.addListener("change", () => {
+        firstListenerRuns += 1;
+      });
+      this.realm.addListener("change", () => {
+        secondListenerRuns += 1;
+      });
+      this.realm.addListener("change", () => {
+        if (firstListenerRuns === 1 && secondListenerRuns === 1) {
+          // After first write, continue
+          return;
+        } else if (firstListenerRuns === 2 && secondListenerRuns === 2) {
+          // After both write, this should be expected state.
+          done();
+        } else {
+          done(Error("did not run other listeners correctly"));
+        }
+      });
+      this.realm.write(() => {
+        this.realm.create("Person", { age: 3, name: "Bob" });
+      });
+      this.realm.write(() => {
+        this.realm.create("Person", { age: 3, name: "Tom" });
+      });
+    });
+    it("should work with multiple listeners of different types", function (this: RealmContext, done) {
+      let changeListenerRuns = 0;
+      let schemaListenerRuns = 0;
+      this.realm.addListener("change", () => {
+        changeListenerRuns += 1;
+      });
+      this.realm.addListener("beforenotify", () => {
+        changeListenerRuns += 1;
```

Comment:
This looks incorrect.

Reply:
that's true... interestingly enough this actually does not affect tests directly. `beforenotify` is ran AFTER the checks so this is still running correctly in terms of checking that both `schema` and `change` are run... `beforenotify` runs after that, so this addition is meaningless. which is a little odd to me, I am still quite confused by what the point of this event is...